sdk/src/azure/iot/az_iot_provisioning_client.c

```diff
@@ -558,3 +561,51 @@ AZ_NODISCARD az_result az_iot_provisioning_client_parse_received_topic_and_paylo
 
   return AZ_OK;
 }
+
+AZ_NODISCARD az_result az_iot_provisioning_client_get_request_payload(
+    az_iot_provisioning_client const* client,
+    az_span custom_payload_property,
+    struct az_iot_provisioning_client_payload_options const* reserved,
+    char* mqtt_payload,
+    size_t mqtt_payload_size,
+    size_t* out_mqtt_payload_length)
+{
+  (void)reserved;
+
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_IS_NULL(reserved);
+  _az_PRECONDITION_NOT_NULL(mqtt_payload);
+  _az_PRECONDITION(mqtt_payload_size > 0);
+
+  az_json_writer xJsonWriter;
+  az_span xBuffer = az_span_create((uint8_t*)mqtt_payload, (int32_t)mqtt_payload_size);
+
+  _az_RETURN_IF_FAILED(az_json_writer_init(&xJsonWriter, xBuffer, NULL));
+  _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(&xJsonWriter));
+  _az_RETURN_IF_FAILED(az_json_writer_append_property_name(&xJsonWriter, prov_registration_id_label));
+  _az_RETURN_IF_FAILED(az_json_writer_append_string(&xJsonWriter, client->_internal.registration_id));
+
+  if  (az_span_size(custom_payload_property) > 0)
+  {
+    _az_RETURN_IF_FAILED(az_json_writer_append_property_name(&xJsonWriter, prov_payload_label));
+    _az_RETURN_IF_FAILED(az_json_writer_append_json_text(&xJsonWriter, custom_payload_property));
+  }
+
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(&xJsonWriter));
+
+  int32_t json_length = az_span_size(az_json_writer_get_bytes_used_in_destination(&xJsonWriter));
+  if (json_length >= mqtt_payload_size)
+  {
+    return AZ_ERROR_NOT_ENOUGH_SPACE;
+  }
+  else
+  {
+    mqtt_payload[json_length] = 0;
+    if (out_mqtt_payload_length != NULL)
+    {
+      *out_mqtt_payload_length = json_length;
+    }
+  }
```

Comment:
Does this need to null terminated? I understand topic being null terminated but payload can be non-null terminated. Whole motivation to have null terminated topic was paho api that expect null terminated topic, but they have no such limitation on payload.


[https://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/pubsync.html](https://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/pubsync.html)


This makes me think that we should take uint8_t * as input to this API. What you think?

Reply:
Yes - this makes a ton of sense.  Even though API review board will need to sign off on a change like this and just generally another look at API, this one makes so much sense I'm just moving now to it.